tools/admin-console/authorizedapps.go

```diff
@@ -0,0 +1,234 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, softwar
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	authorizedappdb "github.com/google/exposure-notifications-server/internal/authorizedapp/database"
+	"github.com/google/exposure-notifications-server/internal/authorizedapp/model"
+	"github.com/google/exposure-notifications-server/internal/serverenv"
+)
+
+type appHandler struct {
+	config *Config
+	env    *serverenv.ServerEnv
+}
+
+func NewAppHandler(c *Config, env *serverenv.ServerEnv) *appHandler {
+	return &appHandler{config: c, env: env}
+}
+
+func (h *appHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "GET" {
+		h.handleGet(w, r)
+		return
+	} else if r.Method == "POST" {
+		h.handlePost(w, r)
+		return
+	}
+
+	m := TemplateMap{}
+	m.AddErrors("Invalid request.")
+	h.config.RenderTemplate(w, "error", m)
+	return
+}
+
+func (h *appHandler) renderError(m TemplateMap, msg string, w http.ResponseWriter) {
+	m.AddErrors(msg)
+	h.config.RenderTemplate(w, "error", m)
+}
+
+func populateAuthorizedApp(a *model.AuthorizedApp, r *http.Request) []string {
+	a.AppPackageName = r.FormValue("AppPackageName")
+	a.Platform = r.FormValue("Platform")
+
+	a.AllowedRegions = make(map[string]struct{})
+	regions := strings.Split(r.FormValue("Regions"), "\n")
+	for _, region := range regions {
+		a.AllowedRegions[region] = struct{}{}
+	}
+
+	var err error
+	errors := []string{}
+	// SafetyNet pieces.
+	a.SafetyNetDisabled, err = strconv.ParseBool(r.FormValue("SafetyNetDisabled"))
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("SafetyNetDisabled, invalid value: %v", err))
+	}
+	a.SafetyNetApkDigestSHA256 = strings.Split(r.FormValue("SafetyNetApkDigestSHA256"), "\n")
+	a.SafetyNetBasicIntegrity, err = strconv.ParseBool(r.FormValue("SafetyNetBasicIntegrity"))
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("SafetyNetBasicIntegrity, invalid value: %v", err))
+	}
+	a.SafetyNetCTSProfileMatch, err = strconv.ParseBool(r.FormValue("SafetyNetCTSProfileMatch"))
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("SafetyNetCTSProfileMatch, invalid value: %v", err))
+	}
+	a.SafetyNetPastTime, err = time.ParseDuration(r.FormValue("SafetyNetPastTime"))
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("SafetyNetPastTime, invalid value: %v", err))
+	}
+	a.SafetyNetFutureTime, err = time.ParseDuration(r.FormValue("SafetyNetFutureTime"))
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("SafetyNetFutureTime, invalid value: %v", err))
```

Comment:
This seems very similar to https://github.com/google/exposure-notifications-server/pull/395/files#diff-37471a2abefb832b04b186b1bbf646bbR91

Reply:
it is subtly different - this is the type conversions from strings, the other one is validating type safe model object properties.